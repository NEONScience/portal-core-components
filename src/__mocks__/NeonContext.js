/**
   NeonContext Mock

   Various tests cover modules that use NeonContext to get state/site/domain metadata. By mocking
   NeonContext and plugging in all fallback data the tested modules will be able to use a hydrated
   NeonContext for more complete testability.

   Usage:
   import 'path/to/src/__mocks__/NeonContext';
*/

import sitesJSON from '../sampleData/sites.json';
import bundlesJSON from '../sampleData/bundles.json';
import statesJSON from '../lib_components/staticJSON/states.json';
import domainsJSON from '../lib_components/staticJSON/domains.json';
import timeSeriesDataProductsJSON from '../lib_components/staticJSON/timeSeriesDataProducts.json';

import NeonContext from '../lib_components/components/NeonContext/NeonContext';
import BundleParser from '../lib_components/parser/BundleParser';

jest.mock('../lib_components/components/NeonContext/NeonContext', () => (
  {
    ...(jest.requireActual('../lib_components/components/NeonContext/NeonContext').default),
    useNeonContextState: jest.fn(),
    FETCH_STATUS: {
      AWAITING_CALL: 'AWAITING_CALL',
      FETCHING: 'FETCHING',
      ERROR: 'ERROR',
      SUCCESS: 'SUCCESS',
    },
  }
));

NeonContext.useNeonContextState.mockReturnValue([
  {
    ...NeonContext.DEFAULT_STATE,
    data: {
      sites: sitesJSON,
      states: statesJSON,
      domains: domainsJSON,
      bundles: BundleParser.parseContext(
        BundleParser.parseBundlesResponse(bundlesJSON),
      ),
      timeSeriesDataProducts: timeSeriesDataProductsJSON,
      stateSites: {}, // derived when sites is fetched, needs to be mocked here?
      domainSites: {}, // derived when sites is fetched, needs to be mocked here?
    },
    isFinal: true,
  },
]);
