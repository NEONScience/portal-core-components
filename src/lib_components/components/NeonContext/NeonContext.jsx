import React, {
  createContext,
  useContext,
  useReducer,
  useEffect,
} from 'react';
import PropTypes from 'prop-types';

import cloneDeep from 'lodash/cloneDeep';

import { of } from 'rxjs';
import { map, catchError } from 'rxjs/operators';

import REMOTE_ASSETS from '../../remoteAssetsMap/remoteAssetsMap';
import AuthService from '../NeonAuth/AuthService';
import NeonGraphQL from '../NeonGraphQL/NeonGraphQL';
import sitesJSON from '../../staticJSON/sites.json';
import statesJSON from '../../staticJSON/states.json';
import domainsJSON from '../../staticJSON/domains.json';
import bundlesJSON from '../../staticJSON/bundles.json';
import timeSeriesDataProductsJSON from '../../staticJSON/timeSeriesDataProducts.json';

const DRUPAL_HEADER_HTML = REMOTE_ASSETS.DRUPAL_HEADER_HTML.KEY;
const DRUPAL_FOOTER_HTML = REMOTE_ASSETS.DRUPAL_FOOTER_HTML.KEY;

export const FETCH_STATUS = {
  AWAITING_CALL: 'AWAITING_CALL',
  FETCHING: 'FETCHING',
  ERROR: 'ERROR',
  SUCCESS: 'SUCCESS',
};

const DEFAULT_STATE = {
  data: {
    sites: {},
    states: statesJSON,
    domains: domainsJSON,
    bundles: bundlesJSON,
    timeSeriesDataProducts: timeSeriesDataProductsJSON,
    stateSites: {}, // derived when sites is fetched
    domainSites: {}, // derived when sites is fetched
  },
  html: {
    [DRUPAL_HEADER_HTML]: null,
    [DRUPAL_FOOTER_HTML]: null,
  },
  fetches: {
    sites: { status: FETCH_STATUS.AWAITING_CALL, error: null },
    auth: { status: null, error: null },
    [DRUPAL_HEADER_HTML]: { status: null, error: null },
    [DRUPAL_FOOTER_HTML]: { status: null, error: null },
  },
  auth: {
    useCore: false,
    isAuthenticated: false,
    isAuthWorking: false,
    isAuthWsConnected: false,
    userData: null,
  },
  isActive: false,
  isFinal: false,
  hasError: false,
};

// Derive values for stateSites and domainSites in state. This is a one-time mapping we
// generate when sites are loaded into state containing lists of site codes for each
// state code / domain code.
const deriveRegionSites = (state) => {
  const stateSites = {};
  const domainSites = {};
  Object.keys(state.data.sites).forEach((siteCode) => {
    const { stateCode, domainCode } = state.data.sites[siteCode];
    if (!stateSites[stateCode]) { stateSites[stateCode] = new Set(); }
    if (!domainSites[domainCode]) { domainSites[domainCode] = new Set(); }
    stateSites[stateCode].add(siteCode);
    domainSites[domainCode].add(siteCode);
  });
  // Fill in empty sets for any states that had no NEON sites
  Object.keys(state.data.states).forEach((stateCode) => {
    if (!stateSites[stateCode]) { stateSites[stateCode] = new Set(); }
  });
  return {
    ...state,
    data: { ...state.data, stateSites, domainSites },
  };
};

/**
   CONTEXT
*/
const Context = createContext(DEFAULT_STATE);

/**
   HOOK
*/
const useNeonContextState = () => {
  const hookResponse = useContext(Context);
  // If called by a component that's not inside a NeonContext the hookResponse will be
  // DEFAULT_STATE. Otherwise it will return an array of length one containing the current state.
  // Thus we double-check here that we got an active state before returning it.
  if (!Array.isArray(hookResponse) || !hookResponse[0].isActive) {
    return [
      { ...DEFAULT_STATE },
      () => {},
    ];
  }
  return hookResponse;
};

/**
   Reducer
*/
const reducer = (state, action) => {
  // Always deep clone fetches as that's the main thing we care about
  // changing to trigger re-renders in the consumer.
  const newState = { ...state, fetches: cloneDeep(state.fetches) };
  const hasValidRemoteAsset = action.asset && Object.keys(REMOTE_ASSETS).includes(action.asset);
  switch (action.type) {
    case 'fetchCalled':
      if (!action.key || !state.fetches[action.key]) { return state; }
      newState.fetches[action.key].status = FETCH_STATUS.FETCHING;
      return newState;

    // Actions for handling sites fetch
    case 'fetchSitesSucceeded':
      newState.fetches.sites.status = FETCH_STATUS.SUCCESS;
      newState.data.sites = action.sites;
      newState.isFinal = true;
      return deriveRegionSites(newState);
    case 'fetchSitesFailed':
      newState.fetches.sites.status = FETCH_STATUS.ERROR;
      newState.fetches.sites.error = action.error;
      newState.isFinal = true;
      newState.hasError = true;
      return newState;

    // Actions for handling auth fetch
    case 'setAuthenticated':
      newState.auth.isAuthenticated = action.isAuthenticated;
      return newState;
    case 'setAuthWorking':
      newState.auth.isAuthWorking = action.isAuthWorking;
      return newState;
    case 'setAuthWsConnected':
      newState.auth.isAuthWsConnected = action.isAuthWsConnected;
      return newState;
    case 'fetchAuthSucceeded':
      newState.fetches.auth.status = FETCH_STATUS.SUCCESS;
      newState.auth.isAuthenticated = !!action.isAuthenticated;
      newState.auth.userData = AuthService.parseUserData(action.response);
      return newState;
    case 'fetchAuthFailed':
      newState.fetches.auth.status = FETCH_STATUS.ERROR;
      newState.auth.isAuthenticated = false;
      newState.auth.userData = null;
      return newState;

    // Actions for handling remote assets
    case 'fetchHtmlSucceeded':
      if (!hasValidRemoteAsset) { return state; }
      newState.fetches[action.asset].status = FETCH_STATUS.SUCCESS;
      newState.html[action.asset] = action.html;
      return newState;
    case 'fetchHtmlFailed':
      if (!hasValidRemoteAsset) { return state; }
      newState.fetches[action.asset].status = FETCH_STATUS.ERROR;
      newState.fetches[action.asset].error = action.error;
      return newState;

    default:
      return state;
  }
};

/**
   Function to convert sites fetch response to the shape we expect
*/
const parseSitesFetchResponse = (sitesArray = []) => {
  if (!Array.isArray(sitesArray)) { return {}; }
  const sitesObj = {};
  sitesArray.forEach((site) => {
    sitesObj[site.siteCode] = {
      siteCode: site.siteCode || site.code,
      description: site.siteDescription || site.description,
      type: site.siteType || site.type,
      stateCode: site.stateCode,
      domainCode: site.domainCode,
      latitude: site.siteLatitude || site.latitude,
      longitude: site.siteLongitude || site.longitude,
      terrain: site.terrain || sitesJSON[site.siteCode].terrain,
      zoom: site.zoom || sitesJSON[site.siteCode].zoom,
    };
  });
  return sitesObj;
};

/**
   Context Provider
*/
const Provider = (props) => {
  const { useCoreAuth, useCoreHeader, children } = props;

  const initialState = cloneDeep(DEFAULT_STATE);
  initialState.isActive = true;
  if (useCoreAuth) {
    initialState.auth.useCore = true;
    initialState.fetches.auth.status = FETCH_STATUS.AWAITING_CALL;
  }
  if (!useCoreHeader) {
    initialState.fetches[DRUPAL_HEADER_HTML].status = FETCH_STATUS.AWAITING_CALL;
    initialState.fetches[DRUPAL_FOOTER_HTML].status = FETCH_STATUS.AWAITING_CALL;
  }
  const [state, dispatch] = useReducer(reducer, initialState);

  // Method to sanitize partial HTML. As delivered presently there are some markup issues that
  // throw warnings when parsed with HTMLReactParser.
  const sanitizePartialHTML = html => html.replace(/ value=""/g, ' initialValue=""');

  // Method to fetch header and/or footer partials
  const fetchPartialHTML = (key) => {
    if (!Object.keys(REMOTE_ASSETS).includes(key)) { return; }
    const { url } = REMOTE_ASSETS[key];
    window.fetch(url, { method: 'GET', headers: { Accept: 'text/html' } })
      .then((res) => {
        if (res.status >= 400) { throw new Error(`${res.status} ${res.statusText}`); }
        return res.text();
      })
      .then((html) => {
        dispatch({ type: 'fetchHtmlSucceeded', asset: key, html: sanitizePartialHTML(html) });
        return of(true);
      })
      .catch((error) => {
        dispatch({ type: 'fetchHtmlFailed', asset: key, error });
        return of(false);
      });
  };

  // Identify any cascading authentication fetches that require
  // the WS to be connected to initiate.
  const cascadeAuthFetches = [];

  // Subject and effect to perform and manage the sites GraphQL fetch
  const fetchMethods = {
    sites: () => {
      NeonGraphQL.getAllSites().pipe(
        map((response) => {
          if (response.response && response.response.data && response.response.data.sites) {
            const sites = parseSitesFetchResponse(response.response.data.sites);
            dispatch({ type: 'fetchSitesSucceeded', sites });
            return of(true);
          }
          dispatch({ type: 'fetchSitesFailed', error: 'malformed response' });
          return of(false);
        }),
        catchError((error) => {
          dispatch({ type: 'fetchSitesFailed', error: error.message });
          return of(false);
        }),
      ).subscribe();
    },
    auth: () => {
      AuthService.fetchUserInfo(
        (response) => {
          const isAuthenticated = AuthService.isAuthenticated(response);
          if (!isAuthenticated && AuthService.isSsoLogin(response)) {
            // If we're not authenticated and have identified another SSO
            // application that's authenticated, trigger a silent authentication
            // check flow.
            if (AuthService.allowSilentAuth()) {
              if (!state.auth.isAuthWsConnected) {
                cascadeAuthFetches.push(() => AuthService.loginSilently(dispatch, true));
              } else {
                AuthService.loginSilently(dispatch, true);
              }
            } else {
              AuthService.login();
            }
          } else {
            dispatch({ type: 'fetchAuthSucceeded', isAuthenticated, response });
          }
          // Initialize a subscription to the auth WS
          AuthService.watchAuth0(dispatch, cascadeAuthFetches);
        },
        (error) => {
          dispatch({ type: 'fetchAuthFailed', error });
          // Initialize a subscription to the auth WS
          AuthService.watchAuth0(dispatch, cascadeAuthFetches);
        },
      );
    },
    [DRUPAL_HEADER_HTML]: () => fetchPartialHTML(DRUPAL_HEADER_HTML),
    [DRUPAL_FOOTER_HTML]: () => fetchPartialHTML(DRUPAL_FOOTER_HTML),
  };

  // Effect: Trigger all fetches that are awaiting call
  useEffect(() => {
    Object.keys(state.fetches).forEach((key) => {
      if (
        state.fetches[key].status !== FETCH_STATUS.AWAITING_CALL
          || typeof fetchMethods[key] !== 'function'
      ) { return; }
      dispatch({ type: 'fetchCalled', key });
      fetchMethods[key]();
    });
  });

  /**
     Render
  */
  return (
    <Context.Provider value={[state, dispatch]}>
      {children}
    </Context.Provider>
  );
};

Provider.propTypes = {
  useCoreAuth: PropTypes.bool,
  useCoreHeader: PropTypes.bool,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.string,
    ])),
    PropTypes.node,
    PropTypes.string,
  ]).isRequired,
};

Provider.defaultProps = {
  useCoreAuth: false,
  useCoreHeader: false,
};

/**
   getWrappedComponent
*/
const getWrappedComponent = Component => (props) => {
  const [{ isActive }] = useNeonContextState();
  if (!isActive) {
    return (
      <Provider>
        <Component {...props} />
      </Provider>
    );
  }
  return <Component {...props} />;
};

/**
   Export
*/
const NeonContext = {
  Provider,
  useNeonContextState,
  DEFAULT_STATE,
  getWrappedComponent,
};
export default NeonContext;
