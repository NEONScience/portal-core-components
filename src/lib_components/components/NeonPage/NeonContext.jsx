import React, {
  createContext,
  useContext,
  useReducer,
  useEffect,
} from 'react';
import PropTypes from 'prop-types';

import { of } from 'rxjs';
import { map, catchError } from 'rxjs/operators';

import NeonGraphQL from '../NeonGraphQL/NeonGraphQL';
import sitesJSON from '../../static/sites/sites.json';
import statesJSON from '../../static/states/states.json';
import domainsJSON from '../../static/domains/domains.json';
import bundlesJSON from '../../static/bundles/bundles.json';

const FETCH_STATUS = {
  AWAITING_CALL: 'AWAITING_CALL',
  FETCHING: 'FETCHING',
  ERROR: 'ERROR',
  SUCCESS: 'SUCCESS',
};

const DEFAULT_STATE = {
  data: {
    sites: {},
    states: statesJSON,
    domains: domainsJSON,
    bundles: bundlesJSON,
  },
  fetches: {
    sites: { status: FETCH_STATUS.AWAITING_CALL, error: null },
  },
};

/**
   CONTEXT
*/
const Context = createContext(DEFAULT_STATE);

/**
   HOOK
*/
const useNeonContextState = () => {
  const hookResponse = useContext(Context);
  if (hookResponse.length !== 2) {
    return [
      { ...DEFAULT_STATE },
      () => {},
    ];
  }
  return hookResponse;
};

/**
   Reducer
*/
const reducer = (state, action) => {
  const newState = { ...state };
  switch (action.type) {
    case 'fetchSitesCalled':
      newState.fetches.sites.status = FETCH_STATUS.FETCHING;
      return newState;
    case 'fetchSitesSucceeded':
      newState.fetches.sites.status = FETCH_STATUS.SUCCESS;
      newState.data.sites = action.sites;
      return newState;
    case 'fetchSitesFailed':
      newState.fetches.sites.status = FETCH_STATUS.ERROR;
      newState.fetches.sites.error = action.error;
      return newState;
    default:
      return state;
  }
};

/**
   Function to convert sites fetch response to the shape we expect
*/
const parseSitesFetchResponse = (sitesArray = []) => {
  if (!Array.isArray(sitesArray)) { return {}; }
  const sitesObj = {};
  sitesArray.forEach((site) => {
    sitesObj[site.siteCode] = {
      siteCode: site.siteCode || site.code,
      description: site.siteDescription || site.description,
      type: site.siteType || site.type,
      stateCode: site.stateCode,
      domainCode: site.domainCode,
      latitude: site.siteLatitude || site.latitude,
      longitude: site.siteLongitude || site.longitude,
      terrain: site.terrain || sitesJSON[site.siteCode].terrain,
    };
  });
  return sitesObj;
};

/**
   Context Provider
*/
const Provider = (props) => {
  const { children } = props;
  const [state, dispatch] = useReducer(reducer, { ...DEFAULT_STATE });

  // Subject and effect to perform and manage the sites GraphQL fetch
  const fetchAllSites$ = NeonGraphQL.getAllSites().pipe(
    map((response) => {
      if (response.response && response.response.data && response.response.data.sites) {
        const sites = parseSitesFetchResponse(response.response.data.sites);
        dispatch({ type: 'fetchSitesSucceeded', sites });
        return of(true);
      }
      dispatch({ type: 'fetchSitesFailed', error: 'malformed response' });
      return of(false);
    }),
    catchError((error) => {
      dispatch({ type: 'fetchSitesFailed', error: error.message });
      return of(false);
    }),
  );
  useEffect(() => {
    if (state.fetches.sites.status === FETCH_STATUS.AWAITING_CALL) {
      dispatch({ type: 'fetchSitesCalled' });
      fetchAllSites$.subscribe();
    }
  });

  /**
     Render
  */
  return (
    <Context.Provider value={[state]}>
      {children}
    </Context.Provider>
  );
};

Provider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.string,
    ])),
    PropTypes.node,
    PropTypes.string,
  ]).isRequired,
};

/**
   Export
*/
const NeonContext = { Provider, useNeonContextState };
export default NeonContext;
