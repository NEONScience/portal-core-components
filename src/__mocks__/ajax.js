/**
   Ajax Mock

   Tests for any components using rjsx/ajax.getJSON can use this to inject a synchronous response.

   Usage:

   import { mockAjaxResponse, mockAjaxError } from 'path/to/src/__mocks__/ajax';

   // Set a mock response (for ajax() AND ajax.getJSON() methods)
   mockAjaxResponse({ ...response });

   // Set a mock error (for ajax() AND ajax.getJSON() methods)
   mockAjaxError('fail');
*/

import { of, throwError } from 'rxjs';

import { ajax } from 'rxjs/ajax';

jest.mock('rxjs/ajax');

ajax.mockReturnValue(of({}));

ajax.getJSON = jest.fn();
ajax.getJSON.mockReturnValue(of({}));

export function mockRawAjaxResponse(response = {}) {
  ajax.mockReset();
  ajax.mockReturnValue(of(response));
}

export function mockRawAjaxError(error = '') {
  ajax.mockReset();
  ajax.mockImplementation(() => throwError(new Error(error)));
}

export function mockGetJSONAjaxResponse(response = {}) {
  ajax.getJSON.mockReset();
  ajax.getJSON.mockReturnValue(of(response));
}

export function mockGetJSONAjaxError(error = '') {
  ajax.getJSON.mockReset();
  ajax.getJSON.mockImplementation(() => throwError(new Error(error)));
}

export function mockAjaxResponse(response = {}) {
  mockRawAjaxResponse(response);
  mockGetJSONAjaxResponse(response);
}

export function mockAjaxError(error = '') {
  mockRawAjaxError(error);
  mockGetJSONAjaxError(error);
}
