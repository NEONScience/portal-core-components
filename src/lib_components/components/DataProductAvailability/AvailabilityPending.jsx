import React from 'react';

import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import CircularProgress from '@material-ui/core/CircularProgress';
import Typography from '@material-ui/core/Typography';

import ErrorIcon from '@material-ui/icons/Warning';

import NeonContext from '../NeonContext/NeonContext';

export default function AvailabilityPending() {
  const [
    { isFinal: neonContextIsFinal, hasError: neonContextHasError },
  ] = NeonContext.useNeonContextState();

  if (neonContextIsFinal && !neonContextHasError) {
    return null;
  }

  const status = (neonContextIsFinal && neonContextHasError) ? 'error' : 'loading';

  return (
    <Card style={{ width: '100%' }}>
      <CardContent
        style={{
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
          padding: '24px',
          flexDirection: 'column',
        }}
      >
        {status === 'loading' ? (
          <React.Fragment>
            <Typography variant="subtitle2" component="h6" gutterBottom>
              Loading Sites...
            </Typography>
            <CircularProgress />
          </React.Fragment>
        ) : (
          <React.Fragment>
            <ErrorIcon fontSize="large" color="error" />
            <Typography variant="subtitle2" component="h6" style={{ marginTop: '8px' }}>
              Error - Unable to Load Sites
            </Typography>
          </React.Fragment>
        )}
      </CardContent>
    </Card>
  );
}
