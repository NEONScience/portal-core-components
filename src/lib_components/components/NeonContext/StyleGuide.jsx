/* eslint-disable react/jsx-one-expression-per-line, jsx-a11y/anchor-is-valid */
import React from 'react';

import { makeStyles } from '@mui/styles';
import Divider from '@mui/material/Divider';
import Link from '@mui/material/Link';
import Typography from '@mui/material/Typography';

import NeonContext from '@/components/NeonContext/NeonContext';
import Theme from '@/components/Theme/Theme';

import DocBlock from '../../../components/DocBlock';
import CodeBlock from '../../../components/CodeBlock';
import ExampleBlock from '../../../components/ExampleBlock';

const useStyles = makeStyles((theme) => ({
  divider: {
    margin: theme.spacing(3, 0),
  },
}));

const NeonContextStateComponent = () => {
  const [{ data, fetches }] = NeonContext.useNeonContextState();
  const {
    sites,
    states,
    domains,
    bundles,
  } = data;
  const preStyle = {
    height: '15vh',
    overflow: 'scroll',
    border: '1px solid black',
    padding: '2px',
  };
  return (
    <div style={{ width: '100%' }}>
      <div>Sites</div>
      <pre style={preStyle}>
        {fetches.sites.status === 'SUCCESS' ? JSON.stringify(sites, null, 2) : fetches.sites.status}
      </pre>
      <div>States</div>
      <pre style={preStyle}>{JSON.stringify(states, null, 2)}</pre>
      <div>Domains</div>
      <pre style={preStyle}>{JSON.stringify(domains, null, 2)}</pre>
      <div>Bundles</div>
      <pre style={preStyle}>
        {fetches.bundles.status === 'SUCCESS' ? JSON.stringify(bundles, null, 2) : fetches.bundles.status}
      </pre>
    </div>
  );
};

export default function StyleGuide() {
  const classes = useStyles(Theme);
  const NeonPageLink = (
    <Link href="#NeonPage">NeonPage</Link>
  );

  return (
    <>

      <DocBlock>
        Many components in Portal Core Components, as well as many apps utilizing Portal Core
        Components (including NeonPage) to stand up functioning pages require a few common and
        infrequently-changing data structures.
      </DocBlock>
      <DocBlock>
        One example of such a data structure is the list of NEON Sites with basic meta data
        (names, types, latitude/longitude, etc.). Also a listing of NEON domains with their codes
        and names, or US States. Some of this information is hard-coded in Portal Core Components
        as JSON files while other information (notably NEON Sites) comes primarily from the server.
      </DocBlock>
      <DocBlock>
        The NeonContext component is a React context that affords a provider and a hook to use its
        state. On mount it will do a one-time fetch and/or import for each of these data structures
        such that any pages or components in the session may get them all from a single source
        without having to worry about fetch management.
      </DocBlock>
      <CodeBlock>
        {`
import NeonContext from 'portal-core-components/lib/components/NeonContext';
        `}
      </CodeBlock>

      <Divider className={classes.divider} />
      <Typography variant="h6" component="h4" gutterBottom>Provider</Typography>

      <DocBlock>
        The NeonContext Provider should be used a the highest order necessary to capture all
        components within that would need to use the state it provides. It takes no props and
        one should never nest a NeonContext Provider within another NeonContext Provider.
      </DocBlock>
      <CodeBlock>
        {`
import NeonContext from 'portal-core-components/lib/components/NeonContext';

const MyComponent = () => {
  ...
  return (
    <NeonContext.Provider>
      ...
    </NeonContext.Provider>
  );
};
        `}
      </CodeBlock>

      <DocBlock>
        <span style={{ fontWeight: 600 }}>
          IMPORTANT NOTE: {NeonPageLink} wraps its childen in a NeonContext Provider. Also any core
          components that require an active context will wrap themselves in one if there is not one
          available. Only use the Provider when wrapping new components that will need to access
          the NeonContext state.
        </span>
      </DocBlock>

      <Divider className={classes.divider} />
      <Typography variant="h6" component="h4" gutterBottom>When Final</Typography>

      <DocBlock>
        Pass a function to the <tt>whenFinal</tt> prop on the Provider to trigger arbitrary logic
        exactly once when the NeonContext is finalized. This can be useful for components or pages
        that rely on NeonContext data to be present and accessible before taking certain actions.
      </DocBlock>
      <DocBlock>
        The <tt>whenFinal</tt> function can no arguments or an single optional argument representing
        a deep copy of the finalized NeonContext state.
      </DocBlock>
      <CodeBlock>
        {`
import NeonContext from 'portal-core-components/lib/components/NeonContext';

const MyComponent = () => {
  ...
  return (
    <NeonContext.Provider
      whenFinal={(finalizedStateClone) => {
        console.log('NeonContext is Ready!', finalizedStateClone);
      }}
    >
      ...
    </NeonContext.Provider>
  );
};
        `}
      </CodeBlock>

      <Divider className={classes.divider} />
      <Typography variant="h6" component="h4" gutterBottom>Authentication</Typography>
      <DocBlock>
        The NeonContext Provider accepts a <tt>useCoreAuth</tt> boolean prop to trigger an auth
        fetch using the core authentication function. This only works for NEON Data Portal pages.
        A future iteration will add support for passing in a custom auth function for third party
        core-components consumers. Until then, if any core components are used in an app deployed
        to the NEON Data Portal but <i>not</i> using <tt>NeonPage</tt>, this is how to trigger
        an auth fetch.
      </DocBlock>
      <CodeBlock>
        {`
import NeonContext from 'portal-core-components/lib/components/NeonContext';

const MyComponent = () => {
  ...
  return (
    <NeonContext.Provider useCoreAuth>
      ...
    </NeonContext.Provider>
  );
};
        `}
      </CodeBlock>

      <Divider className={classes.divider} />
      <Typography variant="h6" component="h4" gutterBottom>Accessing State with useNeonContextState()</Typography>

      <DocBlock>
        NeonContext state can be accessed using the <tt>useNeonContextState()</tt> hook. Note that
        this hook <i>only</i> provides state. It does not provide a dispatch function, in that this
        state cannot be written to by components that consume it.
      </DocBlock>
      <DocBlock>
        The example below deomnstrates how to use the <tt>useNeonContextState()</tt> hook to access
        state in a basic component. This component displays the four NeonContext state data
        structures as they become available:
      </DocBlock>
      <ul>
        <li>
          <b><tt>sites</tt></b> - Meta data object for NEON sites keyed by siteCode
          (e.g. <tt>ABBY</tt>)
        </li>
        <li>
          <b><tt>states</tt></b> - Meta data object for US states keyed by stateCode
          (e.g. <tt>CO</tt>)
        </li>
        <li>
          <b><tt>domains</tt></b> - Meta data object for NEON domains keyed by domainCode
          (e.g. <tt>D01</tt>)
        </li>
        <li>
          <b><tt>bundles</tt></b> - Meta data object containing mappings of children and parents for
          bundled data products
        </li>
      </ul>
      <ExampleBlock>
        <NeonContextStateComponent />
      </ExampleBlock>
      <CodeBlock>
        {`
import NeonPage from 'portal-core-components/lib/components/NeonPage';

const NeonContextStateComponent = () => {
  const [{ data, fetches }] = NeonPage.useNeonContextState();
  const { sites, states, domains, bundles } = data;
  const preStyle = { height: '15vh', overflow: 'scroll', border: '1px solid black', padding: '2px' };
  return (
    <div style={{ width: '100%' }}>
      <div>Sites</div>
      <pre style={preStyle}>
        {fetches.sites.status === 'SUCCESS' ? JSON.stringify(sites, null, 2) : fetches.sites.status}
      </pre>
      <div>States</div>
      <pre style={preStyle}>{JSON.stringify(states, null, 2)}</pre>
      <div>Domains</div>
      <pre style={preStyle}>{JSON.stringify(domains, null, 2)}</pre>
      <div>Bundles</div>
      <pre style={preStyle}>
        {fetches.bundles.status === 'SUCCESS' ? JSON.stringify(bundles, null, 2) : fetches.bundles.status}
      </pre>
    </div>
  );
};

export default NeonContextStateComponent;
        `}
      </CodeBlock>

      <Divider className={classes.divider} />
      <Typography variant="h6" component="h4" gutterBottom>Self-Wrapping Components</Typography>

      <DocBlock>
        Several components in this library require an active NeonContext for the information it
        provides. However these components are also designed to function outside of a NeonPage if
        need be. NeonContext provides <tt>getWrappedComponent()</tt>, a function intended to
        dynamically ensure a NeonContext is there in either use case.
      </DocBlock>
      <CodeBlock>
        {`
import NeonContext from 'portal-core-components/lib/components/NeonContext';
import { resolveProps } from 'portal-core-components/lib/util/defaultProps';

const defaultProps = { ... };

const Foo = (inProps) => {
  const props = resolveProps(defaultProps, inProps);
  const [{ data: neonContextData }] = NeonContext.useNeonContextState();
  ...
  return (...);
}

Foo.propTypes = { ... };

const WrappedFoo = NeonContext.getWrappedComponent(Foo);

export default WrappedFoo;
        `}
      </CodeBlock>
      <DocBlock>
        The example above shows how a component might look when it both consumes NeonContext state
        and self-wraps to make sure such state is always available. The WrappedFoo component,
        exported as default, can still be imported as Foo. Its only purpose is to check if there is
        an active NeonContext and, if not, wrap the component in a Provider.
      </DocBlock>
      <DocBlock>
        Note that since NeonContext generates fetches if two or more self-wrapped components appear
        on a page that does not have a higher-order NeonContext then duplicate siloed fetches will
        happen. This is not very DRY (Don&apos;t Repeat Yourself), so it is incumbent on the
        developer to make add a higher order NeonContext Provider if the situation calls for it.
      </DocBlock>

    </>
  );
}
