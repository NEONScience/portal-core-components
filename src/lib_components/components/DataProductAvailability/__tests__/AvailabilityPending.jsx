import React from 'react';
import renderer from 'react-test-renderer';

import '../../../../__mocks__/NeonContext';
import NeonContext from '../../NeonContext/NeonContext';

import AvailabilityPending from '../AvailabilityPending';

describe('DataProductAvailability - AvailabilityPending', () => {
  beforeEach(() => {
    NeonContext.useNeonContextState.mockReset();
  });
  test('renders correctly for NeonContext that is neither error nor final', () => {
    NeonContext.useNeonContextState.mockReturnValue([{ isFinal: false, hasError: false }]);
    const tree = renderer
      .create(<AvailabilityPending />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('renders correctly for NeonContext that is final with error', () => {
    NeonContext.useNeonContextState.mockReturnValue([{ isFinal: true, hasError: true }]);
    const tree = renderer
      .create(<AvailabilityPending />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('renders correctly for NeonContext that is final without error', () => {
    NeonContext.useNeonContextState.mockReturnValue([{ isFinal: true, hasError: false }]);
    const tree = renderer
      .create(<AvailabilityPending />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
