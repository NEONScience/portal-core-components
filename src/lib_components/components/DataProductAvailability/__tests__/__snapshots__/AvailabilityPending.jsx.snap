// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataProductAvailability - AvailabilityPending renders correctly for NeonContext that is final with error 1`] = `
<div
  className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <div
    className="MuiCardContent-root"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "padding": "24px",
      }
    }
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeLarge"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
      />
    </svg>
    <h6
      className="MuiTypography-root MuiTypography-subtitle2"
      style={
        Object {
          "marginTop": "8px",
        }
      }
    >
      Error - Unable to Load Sites
    </h6>
  </div>
</div>
`;

exports[`DataProductAvailability - AvailabilityPending renders correctly for NeonContext that is final without error 1`] = `null`;

exports[`DataProductAvailability - AvailabilityPending renders correctly for NeonContext that is neither error nor final 1`] = `
<div
  className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <div
    className="MuiCardContent-root"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "padding": "24px",
      }
    }
  >
    <h6
      className="MuiTypography-root MuiTypography-subtitle2 MuiTypography-gutterBottom"
    >
      Loading Sites...
    </h6>
    <div
      className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
      role="progressbar"
      style={
        Object {
          "height": 40,
          "width": 40,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </div>
  </div>
</div>
`;
